File: Java/src/MaximumAverageSubtree.java
public class MaximumAverageSubtree{

    private double res = 0;
    public double maximumAverageSubtree(TreeNode root) {
        dfs(root);
        return res;
    }

    private int[] dfs(TreeNode n){
        if(n == null) return new int [] {0, 0};

        int[] l = dfs(n.left);
        int[] r = dfs(n.right);

        int sum = l[0] + r[0] + n.val;
        int count = l[1] + r[1] + 1;

        res = Math.max(res, (double)sum/count);
        return new int[] {sum, count};
    }

}

File: Python/MaximumAverageSubtree.py
class Solution:
    def maximumAverageSubtree(self, root: TreeNode) -> float:
        self.res = 0
        self.dfs(root)
        return self.res

    def dfs(self, n: TreeNode) -> List[int]:
        if n == None:
            return [0, 0]
        l = self.dfs(n.left)
        r = self.dfs(n.right)

        sum = l[0] + r[0] + n.val
        count = l[1] + r[1] + 1

        self.res = max(self.res, sum / count)
        return [sum, count]

    